# Tidy data with `tidyr`

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE)
```

## What is "tidy" data?

**Tidy** data is where:

* Each **column** describes a **variable**.
* Each **row** describes an **observation**.
* Each **value** is a **cell**.

```{r, echo=FALSE, out.width="100%", fig.cap="Tidy data. [Source](https://garrettgman.github.io/images/tidy-1.png)"}
knitr::include_graphics("images/tidy_data.png")
```


Example of a tidy data:

| day | month  | year  |  weight |  height |
|:---:|:---:|:---:|:---:|:---:|
| 12 | 4  | 2020  |  3.5 |  48 |
| 23 |  8 | 2019  | 2.9  |  50 |
| 9 |  11 | 2020  | 3.8  |  50 |

<br>
Example of untidy data:

| day | month,year  |  weight |  height |
|:---:|:---:|:---:|:---:|
| 12 | 4,2020  |  3.5kg |  48 |
| 23 |  8,2019  | 2.9kg  |  50 |
| 9 |  11,2020  | 3.8kg  |  50 |

Here we introduce some useful functions from the `tidyr` package to clean up and organize data so as to obtain **tidy** data that can then be processed more easily.
<br>



## Separate & Unite

Let's work on the **untidy** tibble **`table5`**, a built-in data set from the `tidyr` package *(table5 displays the number of tuberculosis cases documented by the WHO in selected countries)*:


```{r, echo=F}
knitr::kable(table5, caption ="`table5`")
```

### Separate

Function `separate` separates (indeed!) a column into 2 or more columns, given a specified field separator: `separate(data, col, into, sep)`

```{r}
# separate column "rate" into 2 columns that we name cases and population, based on the "/" separator.
table5_1 <- separate(data=table5, 
         col=rate, 
         into=c("cases", "population"),
         sep="/"
         )

table5_1
```

The default field separator is *a regular expression that matches any sequence of non-alphanumeric values.*
<br>
So this would work well in our example:

```{r}
separate(data=table5, 
         col=rate, 
         into=c("cases", "population")
         )
```

It is good practice to explicitly specify the field separator, so as not to get any bad surprise.


### Unite

Function `unite` does just the opposite: it will unite/stick together 2 or (more) columns: `unite(data, col, column names, sep)`.

```{r}
# unite columns "century" and "year"
table5a_2 <- unite(data=table5_1,
                col=year,
                c("century", "year"), 
                sep="")

table5a_2
```

<center><h4 style="background-color: #a4edff; display: inline-block;">**HANDS-ON**</h4></center>

Fix the following untidy `tibble`:

```{r}
patients <- tibble(
  ID1=c("A", "A", "B", "B"),
  ID2=c(12, 13, 14, 15),
  age=c( 21,  32, 25, 43),
  c("188_93", "167_55", "155_51", "175_72")
)
```

* Separate the two variables that are found in the last column into columns **height** and **weight**: save in object `patients2`.
* Unite columns **ID1** and **ID2** so as to create a unique **ID** per patient: save in `patients3`.

<details>
<summary>
<h5 style="background-color: #a4edff; display: inline-block;">*Answer*</h5>
</summary>

```{r}
# Separate the two variables that are found in the last column.
patients2 <- separate(data=patients, 
         col=4,
         into=c("height", "weight"),
         sep="_")

# Unite columns ID1 and ID2 so as to create a unique ID per patient.
patients3 <- unite(data=patients2,
      col="ID",
      c("ID1", "ID2"),
      sep="")
```

* Is there a problem with `patients3`? Display the tibble in the R console...

Columns `height` and `weight` are **treated as characters** (because they originally contained **underscores** "_" !).
<br> 
We can convert them into numbers with the `convert=TRUE` parameter in `separate`:

```{r}
# convert=TRUE: runs `type.convert` fonction (that converts data to appropriate type)
patients2 <- separate(data=patients, 
                      col=4,
                      into=c("height", "weight"),
                      sep="_", 
                      convert=TRUE)
```

</details>



## Pivoting data

The examples given before were obviously untidy: two values in one cell, date split in two different columns.
<br><br>
It will happen that the format of the data you are usually working with is NOT tidy.

<br>
Example: here we create a tibble that contains the expression of 4 genes in 4 samples: the  gene expression is measured at **two time points** and for **two experimental conditions**:

```{r, eval=T}
df_expr <- tibble(gene=c("Fgfrl1", "Usp34", "Mlf2", "Gga1"),
                KO_t0=1:4,
                KO_t10=5:2,
                WT_t0=6:3,
                WT_t10=0:3
                )
```

**Why this tibble can't be considered `tidy`?**
<br>
* One variable (expression) is spread across 4 columns.

In order to fix this, we can use the `pivot_longer` function: we will **increase the number of rows** and **decrease the number of columns**:

```{r}
pivot_longer(df_expr,
             cols=c("KO_t0", "KO_t10", "WT_t0", "WT_t10")
)
```

Now **only one column** - `value` - contains the **expression** variable.

We can change the name of the newly created columns with parameters `values_to` (where the values are stored) and `names_to` (where the former column names are stored):

```{r}
expr_long <- pivot_longer(df_expr,
             cols=c("KO_t0", "KO_t10", "WT_t0", "WT_t10"),
             values_to="expression",
             names_to="sample"
)
```

Is there anything else that we could do here, to make the data even more **tidy**?

* There are 2 variables in column `sample`: samples **experimental group** (KO or WT) and **time point** (t0 and t10): we can `separate` them.

```{r}
# we can add "remove=FALSE" so as to keep the original column (sample name), that can be useful in this case:
expr_tidy <- separate(expr_long,
              col=sample, 
              into=c("group", "time"),
              sep="_",
              remove=FALSE
         )    
```

The `pivot_wider` function does just the opposite: when an observation is scattered across multiple rows, you can move it back to a column.
<br><br>
Let's take the following example, where expression of 2 genes is measured under **two conditions** (KO and WT) using **two techniques** (RNA-seq and qPCR):
```{r}
df_expr2 <- tibble(gene=c(rep("Fgfrl1", 4), rep("Mlf2", 4)),
                group=rep(rep(c("KO", "WT"), each=2), 2),
                type=rep(c("counts_rnaseq", "Ct_qpcr"), 4),
                value=c(10, 25, 8, 22, 7, 23, 11, 28)
                )
```

Here, it would make sense to have a column `counts_rnaseq` and a column `Ct_qpcr`, which would contain their corresponding values:

```{r}
df_expr2_wider <- pivot_wider(df_expr2,
                        names_from=type, 
                        values_from=value)
```

<center><h4 style="background-color: #a4edff; display: inline-block;">**HANDS-ON**</h4></center>

Make the following data frame **tidy**!

```{r, eval=T}
df_w <- tibble(Name=c("Anna", "Jordi", "Laura", "Miquel"),
                Weight_day0=c(67, 72, 62, 85),
                Weight_day10=c(65, 73, 66, 80),
                )
```

<details>
<summary>
<h5 style="background-color: #a4edff; display: inline-block;">*Answer*</h5>
</summary>

```{r, eval=T}
# `pivot_longer` on "Weight" columns:
df_w2 <- pivot_longer(df_w, 
                    cols=c("Weight_day0", "Weight_day10"),
                    values_to="weight", 
                    names_to="day"
                    )

# `separate` on the "day" column
  # into=c(NA, "day"): the first output is ignored (in the), the "Weight" strings
df_w3 <- separate(df_w2, col = "day", 
                  sep="_", 
                  into=c(NA, "day")
                  )
# If you want to keep only "0" and "10", you can use "day" as a field separator (optionally set convert=TRUE if you want to convert the newly created column to numeric values)
df_w4 <- separate(df_w2, col = "day", 
                  sep="day", 
                  into=c(NA, "day"),
                  convert=TRUE
                  )
# If you want to remove "day" from an existing tibble, you can also use the str_remove function from the stringr package, and optionally add a conversion step:
df_w3$day <- as.numeric(str_remove(df_w3$day, "day"))
```

</details>


                

